File: src/main/java/ru/arc/xserver/XActionMessager.java
package ru.arc.xserver;

import ru.arc.ARC;
import ru.arc.network.ChannelListener;
import ru.arc.util.Common;

import java.util.concurrent.CompletableFuture;

import static ru.arc.util.Logging.error;
import static ru.arc.util.Logging.info;

public class XActionMessager implements ChannelListener {

    public static final String CHANNEL = "arc.xactions";

    @Override
    public void consume(String channel, String message, String originServer) {
        info("[XAction] Received message on channel '{}' from server '{}': {}", channel, originServer, message);
        XAction action;
        try {
            action = Common.gson.fromJson(message, XAction.class);
        } catch (Exception e) {
            error("[XAction] Failed to deserialize action from server '{}': {}", originServer, message, e);
            return;
        }
        if (action == null) {
            error("[XAction] Deserialized null action from server '{}': {}", originServer, message);
            return;
        }
        info("[XAction] Deserialized action type={} from server '{}'", action.getClass().getSimpleName(), originServer);
        XActionManager.run(action);
    }

    public void send(XAction action) {
        CompletableFuture.supplyAsync(() -> {
            String json = Common.gson.toJson(action);
            info("[XAction] Serialized for publish: {}", json);
            return json;
        }).thenAccept(str -> {
            if (ARC.redisManager == null) {
                error("[XAction] Cannot publish — redisManager is null");
                return;
            }
            ARC.redisManager.publish(CHANNEL, str);
            info("[XAction] Published to channel '{}'", CHANNEL);
        }).exceptionally(e -> {
            error("[XAction] Exception during publish", e);
            return null;
        });
    }
}

File: src/main/kotlin/ru/arc/xserver/XActionMessager.kt
package ru.arc.xserver

import ru.arc.redis.RedisOperations
import ru.arc.core.TaskScheduler
import ru.arc.core.Tasks
import ru.arc.util.Common
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info
import ru.arc.util.Logging.withContext
import ru.arc.redis.xaction.TypedRedisBus
import java.util.concurrent.atomic.AtomicBoolean

class XActionMessager(
    redis: RedisOperations,
    private val scheduler: TaskScheduler = Tasks.scheduler,
) : AutoCloseable {

    companion object {
        const val CHANNEL = "arc.xactions"
    }

    private val bus = TypedRedisBus(
        redis = redis,
        channel = CHANNEL,
        gson = Common.gson,
        messageType = XAction::class.java,
        onMessage = { action, originServer ->
            withContext("xaction", null, "receive") {
                info("[XAction] Deserialized action type={} from server '{}'", action.javaClass.simpleName, originServer)
                XActionManager.run(action)
            }
        },
    )
    private val registered = AtomicBoolean(false)
    private val closed = AtomicBoolean(false)

    fun register() {
        check(!closed.get()) { "Cannot register a closed XActionMessager" }
        if (registered.compareAndSet(false, true)) {
            bus.register()
        }
    }

    fun send(action: XAction) {
        if (closed.get()) {
            error("[XAction] Cannot publish through a closed messager")
            return
        }
        scheduler.runAsync {
            if (closed.get()) return@runAsync
            try {
                bus.publish(action)
                info("[XAction] Published to channel '{}'", CHANNEL)
            } catch (e: Exception) {
                error("[XAction] Exception during publish", e)
            }
        }
    }

    override fun close() {
        if (!closed.compareAndSet(false, true)) return
        if (registered.compareAndSet(true, false)) {
            bus.unregister()
        }
    }
}
